mavros/src/plugins/param.cpp

```diff
@@ -48,179 +54,135 @@ using utils::enum_value;
       case enum_value(MT::REAL32):
-        float_tmp = uv.param_float;
-        param_value = float_tmp;
-        break;
-      // [[[end]]] (checksum: 5950e4ee032d4aa198b953f56909e129)
+      	return Parameter(param_name, uv.param_float, pmsg.param_count, pmsg.param_index);
+      // [[[end]]] (checksum: 566aed8be49d83f77c0178b2608a9101)
 
       default:
-        ROS_WARN_NAMED(
-          "param", "PM: Unsupported param %.16s (%u/%u) type: %u",
+        RCLCPP_WARN(
+          rclcpp::get_logger("param"), "PM: Unsupported param %.16s (%u/%u) type: %u",
           pmsg.param_id.data(), pmsg.param_index, pmsg.param_count, pmsg.param_type);
-        param_value = 0;
+        throw std::runtime_error("Unsupported parameter type: " + to_string_enum<MT>(pmsg.param_type));
     }
   }
 
   /**
-   * Variation of set_value with quirks for ArduPilotMega
+   * Variation of from_mavlink with quirks for ArduPilotMega
    */
-  void set_value_apm_quirk(mavlink::common::msg::PARAM_VALUE & pmsg)
+  static Parameter from_mavlink_apm_quirk(mavlink::common::msg::PARAM_VALUE & pmsg)
   {
-    int32_t int_tmp;
-    float float_tmp;
+
+    auto param_name = mavlink::to_string(pmsg.param_id);
 
     switch (pmsg.param_type) {
       // [[[cog:
       // for a, b in param_types:
       //     btype = 'int' if 'int' in b else b
       //     cog.outl("case enum_value(MT::%s):" % a.upper())
-      //     cog.outl("\t%s_tmp = pmsg.param_value;" % btype)
-      //     cog.outl("\tparam_value = %s_tmp;" % btype)
-      //     cog.outl("\tbreak;")
+      //     cog.outl("\treturn Parameter(param_name, pmsg.param_value, pmsg.param_count, pmsg.param_index);")
       // ]]]
       case enum_value(MT::INT8):
-        int_tmp = pmsg.param_value;
-        param_value = int_tmp;
-        break;
-      case enum_value(MT::UINT8):
-        int_tmp = pmsg.param_value;
-        param_value = int_tmp;
-        break;
+      	return Parameter(param_name, pmsg.param_value, pmsg.param_count, pmsg.param_index);
       case enum_value(MT::INT16):
-        int_tmp = pmsg.param_value;
-        param_value = int_tmp;
-        break;
-      case enum_value(MT::UINT16):
-        int_tmp = pmsg.param_value;
-        param_value = int_tmp;
-        break;
+      	return Parameter(param_name, pmsg.param_value, pmsg.param_count, pmsg.param_index);
       case enum_value(MT::INT32):
-        int_tmp = pmsg.param_value;
-        param_value = int_tmp;
-        break;
-      case enum_value(MT::UINT32):
-        int_tmp = pmsg.param_value;
-        param_value = int_tmp;
-        break;
+      	return Parameter(param_name, pmsg.param_value, pmsg.param_count, pmsg.param_index);
       case enum_value(MT::REAL32):
-        float_tmp = pmsg.param_value;
-        param_value = float_tmp;
-        break;
-      // [[[end]]] (checksum: c30ee34dd84213471690612ab49f1f73)
+      	return Parameter(param_name, pmsg.param_value, pmsg.param_count, pmsg.param_index);
```

Comment:
But here you likely get wrong type: pmsg.param_value is float, but it carries bools and integers.

Reply:
I think my plan was to wrap it in an appropriate `type(pmsg.param_value)`  style cast to force the template. Evidently I never got around to it.